nano/rpc/rpc.cpp

```diff
@@ -1,62 +1,50 @@
-#include <boost/algorithm/string.hpp>
-#include <nano/lib/config.hpp>
-#include <nano/lib/interface.h>
-#include <nano/node/node.hpp>
-
+#include <boost/format.hpp>
+#include <nano/lib/rpchandlerinterface.hpp>
 #include <nano/rpc/rpc.hpp>
-#include <nano/rpc/rpc_connection.hpp>
-#include <nano/rpc/rpc_handler.hpp>
+#include <nano/rpc/rpcconnection.hpp>
 
 #ifdef NANO_SECURE_RPC
-#include <nano/rpc/rpc_secure.hpp>
+#include <nano/rpc/rpcsecure.hpp>
 #endif
 
-#include <nano/lib/errors.hpp>
-
-nano::rpc::rpc (boost::asio::io_context & io_ctx_a, nano::node & node_a, nano::rpc_config const & config_a) :
-acceptor (io_ctx_a),
+nano::rpc::rpc (boost::asio::io_context & io_ctx_a, nano::rpc_config const & config_a, nano::rpc_handler_interface & rpc_handler_interface_a) :
 config (config_a),
-node (node_a)
-{
-}
-
-void nano::rpc::add_block_observer ()
+acceptor (io_ctx_a),
+logger (std::chrono::milliseconds (0)),
+io_ctx (io_ctx_a),
+rpc_handler_interface (rpc_handler_interface_a)
 {
-	node.observers.blocks.add ([this](std::shared_ptr<nano::block> block_a, nano::account const & account_a, nano::uint128_t const &, bool) {
-		observer_action (account_a);
-	});
+	rpc_handler_interface.rpc_instance (*this);
 }
 
-void nano::rpc::start (bool rpc_enabled_a)
+nano::rpc::~rpc ()
 {
-	if (rpc_enabled_a)
+	if (!stopped)
 	{
-		auto endpoint (nano::tcp_endpoint (config.address, config.port));
-		acceptor.open (endpoint.protocol ());
-		acceptor.set_option (boost::asio::ip::tcp::acceptor::reuse_address (true));
-
-		boost::system::error_code ec;
-		acceptor.bind (endpoint, ec);
-		if (ec)
-		{
-			node.logger.always_log (boost::str (boost::format ("Error while binding for RPC on port %1%: %2%") % endpoint.port () % ec.message ()));
-			throw std::runtime_error (ec.message ());
-		}
-
-		acceptor.listen ();
+		stop ();
 	}
+}
 
-	add_block_observer ();
+void nano::rpc::start ()
+{
+	auto endpoint (boost::asio::ip::tcp::endpoint (config.address, config.port));
+	acceptor.open (endpoint.protocol ());
+	acceptor.set_option (boost::asio::ip::tcp::acceptor::reuse_address (true));
 
-	if (rpc_enabled_a)
+	boost::system::error_code ec;
+	acceptor.bind (endpoint, ec);
+	if (ec)
 	{
-		accept ();
+		logger.always_log (boost::str (boost::format ("Error while binding for RPC on port %1%: %2%") % endpoint.port () % ec.message ()));
+		throw std::runtime_error (ec.message ());
 	}
+	acceptor.listen ();
+	accept ();
 }
 
 void nano::rpc::accept ()
 {
-	auto connection (std::make_shared<nano::rpc_connection> (node, *this));
+	auto connection (std::make_shared<nano::rpc_connection> (config, network_constants, io_ctx, logger, rpc_handler_interface));
 	acceptor.async_accept (connection->socket, [this, connection](boost::system::error_code const & ec) {
```

Comment:
Ideally the external RPC process should establish persistent connection with its HTTP clients to reduce file descriptor use and improve performance. Now it's one connection per call (i.e same as before). 

One of the benefits of IPC is persistent connections and it would be nice if that extended to http clients. Not sure if such a change is in scope for v19 though.

Reply:
Not sure it is possible to maintain connections with HTTP as it is stateless, but I think it would be the responsibility for another PR regardless (> v19).